### PR TITLE
RFC: content addressed file system

### PR DESCRIPTION
**Pull request by b5**
first draft to get cafs into the rfc process.

### CODE REVIEW

File: text/0003-content_addressed_file_system.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+- RFC PR: [#3](https://github.com/qri-io/rfcs/pull/3)
+- Issue: NA
+
+# Summary
+[summary]: #summary
+
+Content-Addressed File System (cafs) is a generalized interface for working with 
+filestores that determine names content based on the content itself, usually
+through some sort of hashing function.
+Examples of a content-addressed file systems include git, bittorrent, IPFS,
```
Comment (rgardaphe) [inline text/0003-content_addressed_file_system.md]: should read: "Examples of content-addressed file systems include" (take out the 'a')

File: text/0003-content_addressed_file_system.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+subset of features qri needs to function.
+
+This package doesn't aim to implement everything a given filestore can do, 
+but instead focus on basic file & directory i/o. cafs is very early days, 
+starting with a proof of concept based on IPFS and an in-memory implementation. 
+Over time we'll work to add additional stores, which will undoubtably affect 
+the overall interface definition.
+
+A tacit goal of this interface is to manage the seam between graph-based
+storage systems, and and a file interface.
```
Comment (rgardaphe) [inline text/0003-content_addressed_file_system.md]: Should read: "storage systems, and a file interface." (cut one 'and')

File: text/0003-content_addressed_file_system.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+[motivation]: #motivation
+
+The long-term goal of cafs is to define an interface for common filestore 
+operations between different content-addressed filestores that serves the
+subset of features qri needs to function.
+
+This package doesn't aim to implement everything a given filestore can do, 
+but instead focus on basic file & directory i/o. cafs is very early days, 
+starting with a proof of concept based on IPFS and an in-memory implementation. 
+Over time we'll work to add additional stores, which will undoubtably affect
```
Comment (rgardaphe) [inline text/0003-content_addressed_file_system.md]: Should read: "which will undoubtedly affect"

File: text/0003-content_addressed_file_system.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+# Motivation
+[motivation]: #motivation
+
+The long-term goal of CAFS is to define an interface for common filestore 
+operations between different content-addressed filestores that serves the
+subset of features qri needs to function.
+
+This package doesn't aim to implement everything a given filestore can do, 
+but instead focus on basic file & directory i/o. CAFS is in its very early days, 
+starting with a proof of concept based on IPFS and an in-memory implementation.
```
[the remark this answers is not in the document]
Comment (b5) [reply text/0003-content_addressed_file_system.md]: I'd rather leave the "next" implementation out of the RFC for now b/c we don't have any plans to write another CAFS implementation, but I'd say:
- Git
- Dat
- Our own implementation built on a key/value datastore

in that order

File: text/0003-content_addressed_file_system.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  // This interface is under active development, expect it to change lots.
+  // It's currently form-fitting around IPFS (ipfs.io), with far-off plans to generalize
+  // toward compatibility with git (git-scm.com), then maybe other stuff, who knows.
+  type Filestore interface {
+    // Put places a file or a directory in the store.
+    // The most notable difference from a standard file store is the store itself determines
+    // the resulting key (google "content addressing" for more info ;)
+    // keys returned by put must be prefixed with the PathPrefix,
+    // eg. /ipfs/QmZ3KfGaSrb3cnTriJbddCzG7hwQi2j6km7Xe7hVpnsW5S
+    Put(file File, pin bool) (key datastore.Key, err error)
```
Comment (dustmop) [inline text/0003-content_addressed_file_system.md]: We never define `datastore`. Mention at least a stub definition up above.

`pin` isn't mentioned here either. Should have at least mention a quick explanation of what the intent is.

File: text/0003-content_addressed_file_system.md [hunk 6]
```diff
@@ -11,35 +11,35 @@ we'd hope, or less complete than we'd expect from a new RFC.
 # Summary
 [summary]: #summary
 
-Content-Addressed File System (cafs) is a generalized interface for working with 
-filestores that determine names content based on the content itself, usually
+Content-Addressed File System (CAFS) is a generalized interface for working with 
+filestores that names content based on the content itself, usually
 through some sort of hashing function.
-Examples of a content-addressed file systems include git, bittorrent, IPFS, 
+Examples of content-addressed file systems include include git, bittorrent, IPFS,
```
Comment (dustmop) [inline text/0003-content_addressed_file_system.md]: Two "includes"

File: text/0003-content_addressed_file_system.md (hunk 5 above)
Comment (b5) [reply text/0003-content_addressed_file_system.md]: Ah nice catch. Regarding "datastore", I'm going to change that to raw string values instead, b/c I don't think we need to talk about the datastore interface at all here (I think it'll only add confusion, and is specific to IPFS).

Re pin, will def add a comment

**Comment by ramfox** [comment]
🔒 🔑 🗄

**Comment by b5** [comment]
👯‍♂️ 👯‍♂️ 👯‍♂️ 👯‍♂️ 👯‍♂️ 👯‍♂️ 👯‍♂️ 👯‍♂️ 👯‍♂️ 👯‍♂️ 👯‍♂️ 👯‍♂️

**Pull request merged**